fastly/block_fastly_service_v1_logging_openstack.go

```diff
@@ -0,0 +1,289 @@
+		opts := buildCreateOpenstack(lf, serviceID, latestVersion)
+
+		log.Printf("[DEBUG] Fastly OpenStack logging addition opts: %#v", opts)
+
+		if err := createOpenstack(conn, opts); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h *OpenstackServiceAttributeHandler) Read(d *schema.ResourceData, s *gofastly.ServiceDetail, conn *gofastly.Client) error {
+	// Refresh OpenStack.
+	log.Printf("[DEBUG] Refreshing OpenStack logging endpoints for (%s)", d.Id())
+	openstackList, err := conn.ListOpenstack(&gofastly.ListOpenstackInput{
+		Service: d.Id(),
+		Version: s.ActiveVersion.Number,
+	})
+
+	if err != nil {
+		return fmt.Errorf("[ERR] Error looking up OpenStack logging endpoints for (%s), version (%v): %s", d.Id(), s.ActiveVersion.Number, err)
+	}
+
+	ell := flattenOpenstack(openstackList)
+
+	if err := d.Set(h.GetKey(), ell); err != nil {
+		log.Printf("[WARN] Error setting OpenStack logging endpoints for (%s): %s", d.Id(), err)
+	}
+
+	return nil
+}
+
+func createOpenstack(conn *gofastly.Client, i *gofastly.CreateOpenstackInput) error {
+	_, err := conn.CreateOpenstack(i)
+	return err
+}
+
+func deleteOpenstack(conn *gofastly.Client, i *gofastly.DeleteOpenstackInput) error {
+	err := conn.DeleteOpenstack(i)
+
+	errRes, ok := err.(*gofastly.HTTPError)
+	if !ok {
+		return err
+	}
+
+	// 404 response codes don't result in an error propagating because a 404 could
+	// indicate that a resource was deleted elsewhere.
+	if !errRes.IsNotFound() {
+		return err
+	}
+
+	return nil
+}
+
+func flattenOpenstack(openstackList []*gofastly.Openstack) []map[string]interface{} {
+	var lsl []map[string]interface{}
+	for _, ll := range openstackList {
+		// Convert OpenStack logging to a map for saving to state.
+		nll := map[string]interface{}{
+			"name":               ll.Name,
+			"url":                ll.URL,
+			"user":               ll.User,
+			"bucket_name":        ll.BucketName,
+			"access_key":         ll.AccessKey,
+			"public_key":         ll.PublicKey,
+			"gzip_level":         ll.GzipLevel,
+			"message_type":       ll.MessageType,
+			"path":               ll.Path,
+			"period":             ll.Period,
+			"timestamp_format":   ll.TimestampFormat,
+			"format":             ll.Format,
+			"format_version":     ll.FormatVersion,
+			"placement":          ll.Placement,
+			"response_condition": ll.ResponseCondition,
+		}
+
+		// Prune any empty values that come from the default string value in structs.
+		for k, v := range nll {
+			if v == "" {
+				delete(nll, k)
+			}
+		}
+
+		lsl = append(lsl, nll)
+	}
+
+	return lsl
+}
+
+func buildCreateOpenstack(openstackMap interface{}, serviceID string, serviceVersion int) *gofastly.CreateOpenstackInput {
+	df := openstackMap.(map[string]interface{})
+
+	return &gofastly.CreateOpenstackInput{
+		Service:           serviceID,
+		Version:           serviceVersion,
+		Name:              gofastly.NullString(df["name"].(string)),
+		URL:               gofastly.NullString(df["url"].(string)),
```

Comment:
Just noting that this URL is used for authentication. The name `url` aligns with the API and client library.